app/auth/authV2/actions/AgentHasClientDetails.scala

```diff
@@ -18,41 +18,68 @@ package auth.authV2.actions
 
 import auth.authV2.AuthExceptions._
 import auth.authV2.EnroledUser
+import config.FrontendAppConfig
 import controllers.agent.routes
 import controllers.agent.sessionUtils.SessionKeys
+import play.api.Logger
 import play.api.mvc.Results.Redirect
 import play.api.mvc.{ActionRefiner, Result}
+import services.SessionDataService
 import uk.gov.hmrc.auth.core.AffinityGroup.Agent
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
 
 import javax.inject.Inject
 import scala.concurrent.{ExecutionContext, Future}
 
-class AgentHasClientDetails @Inject()(implicit val executionContext: ExecutionContext) extends ActionRefiner[EnroledUser, EnroledUser] {
+class AgentHasClientDetails @Inject()(implicit val executionContext: ExecutionContext,
+                                      sessionDataService: SessionDataService, appConfig: FrontendAppConfig)
+  extends ActionRefiner[EnroledUser, EnroledUser] {
 
   lazy val noClientDetailsRoute: Result = Redirect(routes.EnterClientsUTRController.show)
 
   override protected def refine[A](request: EnroledUser[A]): Future[Either[Result, EnroledUser[A]]] = {
 
     val hasConfirmedClient: Boolean = request.session.get(SessionKeys.confirmedClient).nonEmpty
 
-    val hasClientDetails: Boolean = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter
+      .fromRequestAndSession(request, request.session)
+
+    val cookiesHaveClientDetails: Boolean = {
```

Comment:
might need to changes it to lazy in order to postpone this expression evaluation;